README

This is README file for transproc python script and other related scripts.
The purpose of this script is to connect to data source (IMAP, HTTP...), download
new data and process them. Processing is done by dedicated scripts,
where each script handles one transcript format. It is possible to specify
mapping between the transcript processor and data source by several attributes
(like email attribute FROM). Based on these attributes, data is processed by
appropriate processor.

There is a configuration file for transproc, see example transproc.conf
in repository. The configuration file is searched in /etc/fred directory
by default. It is possible to specify its location on command line.


ChangeLog
2019-11-20 (1.7.1)

    * Update spec file for F31 and Centos/RHEL 8

2019-03-20 (1.7.0)

    * License GNU GPLv3+

    * Fix rpm build

2019-02-05 (1.6.1)

    * Change severity of console logger (INFO -> ERROR)

2019-01-15 (1.6.0)

    * Add CSOB CEB processor (SOAP + REST)

2018-10-26 (1.5.0)

    * Generalize backend command configuration
        * backend_payment_cmd
        * backend_statement_cmd (see command placeholders in example configuration)

    * Use setuptools for packaging

    * COPR build

    * Removed obsolete processor proc_ebanka.py

2018-02-14 (1.4.0)

    * Add support for dynamic date interval in url for http sources
        * url placeholders {START_DATE} {END_DATE}
        * see example transproc.conf for usage

2017-03-02 (1.3.0)

    * Configuration file documentation

    * Fedora packaging

2013-06-12 Michal Strnad (1.2.1)

    * Parser for FIO updated to new format

2013-01-07 Michal Strnad (1.2.0)

    * Parser for FIO xml data files

2012-09-10 Tomas Divis, Vlastimil Zima (1.1.2)

    * Whitespace normalization and PEP8-ification

    * Update due to distutils changes (setup.cfg)

2012-06-09 Tomas Divis, Jiri Sadek (1.1.1)

    * fix - escape function for output xml data (xml entities issue)

2012-05-14 Tomas Divis (1.1.0)

    * add support for IMAP SSL

2012-03-19 Jiri Sadek (1.0.13)

    * setup.py fix - config file path

2012-03-15 Jiri Sadek (1.0.12)

    * config option for change of maximal input read iterations of backend
      processing command

    * ebanka csv processor now picks only realized payments for further processing

2010-09-11 Tomas Divis, Jiri Sadek (1.0.11)

    * config fix - imapfolder parameter - preserve default value
      of select(...) - INBOX

2010-08-13 Jiri Sadek (1.0.10)

    * payments in foreign currency are now included in output
      xml (with price set to 0)

2010-06-22 Jiri Sadek (1.0.9)

    * imapfolder config option for IMAP data source

    * default log path using installation prefix now

2010-04-01 Tomas Divis (1.0.5)

    * Added support for logging

    * BugFix: When emtpy csv comes to proc_csob_xml processor,
      it will not generate invalid XML but rather empty string

    * Transproc now don't send empty output XML from processor to backend
      (due to bugfix above)

2010-03-12 Tomas Divis (1.0.2)

    * Setup.py enhanced

    * Some minor bugfixes like exception handling while connecting to HTTP,
      typos...

2010-02-19 Tomas Divis (1.0.0)

    * Support for IMAP + HTTP data sources

    * Support for data source types options

    * Configurable source data recoding from different charset

    * Configurable filter by subject for IMAP source

    * Processed mail marking improved

    * Import command takes parameter for original statement file and its
    mimetype (defined in configuration)

    * Output xml changes:
            * xml header added
            * oldBalance element renamed to old_balance
            * account_bank_code added to statement xml (account_number
            splitted)
            * ident should be same if there is more sources for one account
            (it was set to empty string in proc_ebanka for now)
            * status element added

    * proc_ebanka is now not supported because of 'ident' element problem

2009-02-11 Ales Dolezal (1.0.0)

    * Parser for CSOB xml files.

2007-11-07 Jan Kryl (version 1.0.0)

    * Old get_transcripts.py script imported and renamed to transproc.

    * Parser for ebanka transcript completed.

    * New setup.py installer.


proc_ebanka.py
#!/usr/bin/env python
#

import sys
from fred_transproc.template import template_head
from fred_transproc.template import template_tail
from fred_transproc.template import template_item

def error(msg):
    sys.stderr.write('Invalid format: ' + msg + '\n')
    sys.exit(1)

def getfield(str, i, delim=' '):
    list = [ field for field in str.split(delim) if field ]
    if len(list) < i:
        error('Required field %d is not present in string "%s"' % (i, str))
    return list[i - 1]


if __name__ == '__main__':

    sys.stderr.write('''proc_ebanka processor is not up to date!!!
    There were several changes in format of output XML which are not possible
    to do for text source of statement from raifaisen bank (ebanka)
    like unique identifier of payment, so this processor is obsolete.''')
    sys.exit(1)

    input = sys.stdin.read()
    input = input.replace('\r', '').strip() # delete \r characters
    sections = [ section.strip() for section in input.split('\n\n') if section ]
    if len(sections) != 3:
        if len(sections) >= 4 and not sections[3].startswith('Detaily ZPS'):
            error('Transcript has %d sections and should have 3' % len(sections))

    # header 1
    lines = [ line.strip() for line in sections[0].split('\n') ]
    if not lines[0] in ('Raiffeisenbank a.s.', 'eBanka'):
        error('Transcript does not start with "Raiffeisenbank a.s." word')
    var_number = getfield(lines[1], 4)
    if lines[2].find('/') != -1:
        dates = getfield(lines[2], 3).split('/')
        tmp_date_old = dates[0].split('.')
        tmp_date_new = dates[1].split('.')
        var_date_new = tmp_date_new[2] + '-' + tmp_date_new[1] + '-' + tmp_date_new[0]
        var_date_old = tmp_date_old[2] + '-' + tmp_date_old[1] + '-' + tmp_date_old[0]
    else:
        tmp_date = getfield(lines[2], 2).split('.')
        var_date_new = tmp_date[2] + '-' + tmp_date[1] + '-' + tmp_date[0]
        var_date_old = var_date_new
    year = var_date_new.split('-')[0][2:]
    # header 2
    lines = [ line.strip() for line in sections[1].split('\n') ]
    var_account_id = getfield(lines[1], 3)
    # header 3 + body
    subsections = [ section.strip('\n') for section in sections[2].split('=' * 86) if section ]
    if len(subsections) not in [2, 5]:
        error('Transcript has %d subsections and should have 3 or 5' %
                len(subsections))
        # header 3
    lines = [ line.strip() for line in subsections[1].split('\n') ]
    var_oldBalance = getfield(lines[0], 2, '  ').replace(' ', '')
    var_credit = getfield(lines[1], 3, '  ').replace(' ', '')
    var_debet = getfield(lines[2], 3, '  ').replace(' ', '')
    var_balance = getfield(lines[4], 2, '  ').replace(' ', '')
    # print what we have so far
    print template_head % (var_account_id, var_number, var_date_new, var_balance,
            var_date_old, var_oldBalance, var_credit, var_debet)
    # body
    if len(subsections) > 2:
        transfers = [ transfer for transfer in subsections[4].split('-' * 86) if transfer ]
        for transfer in transfers:
            lines = [ line for line in transfer.split('\n') if line ]#[:3]
            if not (len(lines) == 3 or len(lines) == 4 or len(lines) == 5):
                error('Bad number of lines of transaction item')
            item_spec_symbol = lines[0][44:55].strip()
            item_number = lines[0][:5].strip()
            item_price = lines[0][55:76].strip().replace(' ', '')
            if item_price == '':
                item_price = lines[0][79:86].strip().replace(' ', '')
            item_date_tmp = lines[0][5:11].strip().strip('.').split('.')
            item_date = '20' + year + '-' + item_date_tmp[1] + '-' + item_date_tmp[0]
            item_time = lines[1][5:11].strip()
            item_name = lines[1][11:33].strip()
            item_currency = lines[1][34:36].strip()
            item_var_symbol = lines[1][44:55].strip()
            item_const_symbol = lines[2][44:55].strip()
            if item_price.startswith('-'):
                item_code = '1'
            else:
                item_code = '2'

            # payments between us and bank (and vice versa) has constant symbol 598 or 898
            if item_const_symbol == '598' or item_const_symbol == '898':
                item_type = '2'
            else:
                # other transfers are from registrars
                item_type = '1'

            if lines[2].find('/') != -1:
                (item_account_number, item_account_bank_code) = \
                        lines[2][11:33].strip().split('/')
            else:
                item_account_number, item_account_bank_code = \
                        var_account_id.split('/')

            if len(lines) == 4:
                item_memo = lines[3].strip()
            elif len(lines) == 5:
                # this is for foreign currency transfers
                item_memo = lines[3].strip() + lines[4].strip()
            else:
                item_memo = lines[0][11:33].strip()

            # there is a limit in the database (varchar(64))
            item_memo = item_memo[:63]
            item_ident = '' # this statement does not contain ident as ebanka_csv source :(

            print template_item % (item_ident,
                    item_account_number, item_account_bank_code,
                    item_const_symbol, item_var_symbol, item_spec_symbol,
                    item_price, item_type, item_code, item_memo, item_date,
                    item_date + ' ' + item_time, item_name)

    print template_tail
    sys.exit(0)


proc_fio_xml.py
#! /usr/bin/env python2
#
# Copyright (C) 2012-2019  CZ.NIC, z. s. p. o.
#
# This file is part of FRED.
#
# FRED is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# FRED is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with FRED.  If not, see <https://www.gnu.org/licenses/>.

# according to http://www.fio.cz/docs/cz/API_Bankovnictvi.pdf
# https://www.fio.cz/ib_api/rest/periods/{token}/2012-01-01/2013-12-31/transactions.xml

import sys
import xml.etree.ElementTree
import re
from datetime import datetime

from fred_transproc.template import render_template_head, render_template_tail, render_template_item

# FIO date: YYYY-mm-dd+GMT
# GMT: HH:MM
re_date = re.compile("^([0-9]+-[0-9]+-[0-9]+)[+-][0-9]+:[0-9]+$")
re_xml_tag = re.compile("^({[^}]*})?(.*)$")

def get_isoformat_date(fio_date):

    global re_date
    if not fio_date:
        return ""
    re_fio_date = re_date.match(fio_date)
    if not re_fio_date:
        return ""
    return datetime.strptime(re_fio_date.group(1), "%Y-%m-%d").date().isoformat()


def parse_tag(tag):

    global re_xml_tag
    r_tag = re_xml_tag.match(tag)
    if r_tag is not None:
        return {'ns':r_tag.group(1), 'tag':r_tag.group(2)}
    return {'ns':'', 'tag':tag}


if __name__ == '__main__':
    top = xml.etree.ElementTree.ElementTree()
    root = top.parse(sys.stdin)

    tag = parse_tag(root.tag)
    if tag['tag']!="AccountStatement":
        sys.exit(1)
    nsFio = tag['ns']
    if nsFio is None:
        nsFio = ""
    info = root.find(nsFio+"Info")
    if info is None:
        sys.exit(1)

    info_data = {'accountId':"",
                 'bankId':"",
                 'idList':"",
                 'dateStart':"",
                 'openingBalance':"",
                 'dateEnd':"", 
                 'closingBalance':"",
                 'credit':"",
                 'debet':""}

    for item in info:
        tag = parse_tag(item.tag)['tag']
        if tag in ('dateStart', 'dateEnd'):
            info_data[tag] = get_isoformat_date(item.text)
        elif tag in info_data:
            info_data[tag] = item.text

    print unicode(render_template_head([info_data['accountId'], info_data['bankId'], info_data['idList'],
                                        info_data['dateEnd'], info_data['closingBalance'],
                                        info_data['dateStart'], info_data['openingBalance'],
                                        info_data['credit'], info_data['debet']])).encode('utf8')

    transaction_list = root.find(nsFio+"TransactionList")
    if transaction_list is not None:
        tag_ident             = nsFio + "column_22"
        tag_account_number    = nsFio + "column_2"
        tag_account_bank_code = nsFio + "column_3"
        tag_const_symbol      = nsFio + "column_4"
        tag_var_symbol        = nsFio + "column_5"
        tag_spec_symbol       = nsFio + "column_6"
        tag_currency          = nsFio + "column_14"
        tag_price             = nsFio + "column_1"
        tag_name              = nsFio + "column_10"
        tag_date              = nsFio + "column_0"
        tag_memo              = nsFio + "column_16"
        for tr in transaction_list:
            if parse_tag(tr.tag)['tag']!="Transaction":
                continue
            ident = tr.findtext(tag_ident)
            account_number = tr.findtext(tag_account_number, "")
            account_bank_code = tr.findtext(tag_account_bank_code, "")
            const_symbol = tr.findtext(tag_const_symbol, "")
            var_symbol = tr.findtext(tag_var_symbol, "")
            spec_symbol = tr.findtext(tag_spec_symbol, "")
            currency = tr.findtext(tag_currency, "")
            if currency.upper()=="CZK":
                price = tr.findtext(tag_price, "0").replace("+", "").replace(",", ".")
            else:
                price = "0"
            name = tr.findtext(tag_name, "")
            date = get_isoformat_date(tr.findtext(tag_date))
            code = "1" # 1-normal transaction; 2-storno transaction
            memo = tr.findtext(tag_memo, "")[:64]
            crtime = '' # time when we inserted it to our database (transproc don't fill it)

            type = "1"  # not decided (not processed)

            # all payments in FIO XML are realized:
            status = "1" # 1-Realized (only this should be further processed)

            print unicode(render_template_item([ident, account_number, account_bank_code,
                    const_symbol, var_symbol, spec_symbol, price, type, code, status, memo,
                    date, crtime, name])).encode('utf8')

    print unicode(render_template_tail()).encode('utf8')


proc_csob_xml.py
#! /usr/bin/env python2
#
# Copyright (C) 2009-2019  CZ.NIC, z. s. p. o.
#
# This file is part of FRED.
#
# FRED is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# FRED is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with FRED.  If not, see <https://www.gnu.org/licenses/>.

# according to http://www.csob.cz/cz/Produktovy-katalog/Elektronicke-bankovnictvi/CSOB-BusinessBanking-24/Stranky/Prirucky-a-navody-CSOB-BusinessBanking-24.aspx

import sys
import xml.etree.ElementTree
from datetime import datetime

from fred_transproc.template import render_template_head, render_template_tail, render_template_item


if __name__ == '__main__':
    top = xml.etree.ElementTree.ElementTree()
    top.parse(sys.stdin)
    finsta03 = top.find("FINSTA03")
    account_number_our, account_bank_code_our = finsta03.findtext("S25_CISLO_UCTU").split('/')
    number = finsta03.findtext("S28_CISLO_VYPISU")

    old_date = datetime.strptime(finsta03.findtext("S60_DATUM"), "%d.%m.%Y").date().isoformat()
    old_balance = finsta03.findtext("S60_CASTKA").replace(",", ".")
    old_balance_sign = finsta03.findtext("S60_CD_INDIK").replace("C", "").replace("D", "-")

    date = datetime.strptime(finsta03.findtext("S62_DATUM"), "%d.%m.%Y").date().isoformat()
    balance = finsta03.findtext("S62_CASTKA").replace(",", ".")
    balance_sign = finsta03.findtext("S62_CD_INDIK").replace("C", "").replace("D", "-")

    credit = finsta03.findtext("SUMA_KREDIT", "").replace(",", ".").replace("=", "")
    debet = finsta03.findtext("SUMA_DEBIT", "").replace(",", ".").replace("=", "")

    old_balance = old_balance_sign + old_balance
    balance = balance_sign + balance
    credit = credit
    debet = debet

    print unicode(render_template_head([account_number_our, account_bank_code_our, number, date, balance, old_date,
            old_balance, credit, debet])).encode('utf8')

    finsta05 = finsta03.findall("FINSTA05")
    for item in finsta05:
        ident = item.findtext("S28_POR_CISLO")
        account_number = item.findtext("PART_ACCNO", "")
        account_bank_code = item.findtext("PART_BANK_ID", "")
        const_symbol = item.findtext("S86_KONSTSYM", "")
        var_symbol = item.findtext("S86_VARSYMOUR", "")
        spec_symbol = item.findtext("S86_SPECSYMOUR", "")
        price = item.findtext("S61_CASTKA").replace("+", "").replace(",", ".")
        name = item.findtext("PART_ACC_ID")
        date = datetime.strptime(item.findtext("DPROCD"), "%d.%m.%Y").date().isoformat()
        code = item.findtext("S61_CD_INDIK")
        memo = item.findtext("PART_ID1_1")[:64]
        crtime = ''

        # code can be D - debet, C - credit, DR - storno deposit and CR - storno credit
        # (we already know difference between C and D by sign of price - so to <code> tag
        #  we save only whether it is normal (1) or storno(2) transaction):
        if code in ("D", "C"):
            code = "1"
        elif code in ("DR", "CR"):
            code = "2"

        type = "1"  # not decided (not processed)

        # all payments in CSOB XML are realized:
        status = "1"

        print unicode(render_template_item([ident, account_number, account_bank_code,
                const_symbol, var_symbol, spec_symbol, price, type, code, status, memo,
                date, crtime, name])).encode('utf8')

    print unicode(render_template_tail()).encode('utf8')


fred_transproc/template.py
#
# Copyright (C) 2009-2019  CZ.NIC, z. s. p. o.
#
# This file is part of FRED.
#
# FRED is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# FRED is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with FRED.  If not, see <https://www.gnu.org/licenses/>.

from xml.sax.saxutils import escape
import types

_template_head = '''<?xml version="1.0" encoding="UTF-8"?>
<statements>
  <statement>
    <account_number>%s</account_number>
    <account_bank_code>%s</account_bank_code>
    <number>%s</number>
    <date>%s</date>
    <balance>%s</balance>
    <old_date>%s</old_date>
    <old_balance>%s</old_balance>
    <credit>%s</credit>
    <debet>%s</debet>
    <items>'''

_template_tail = '''    </items>
  </statement>
</statements>'''

_template_item = '''      <item>
        <ident>%s</ident>
        <account_number>%s</account_number>
        <account_bank_code>%s</account_bank_code>
        <const_symbol>%s</const_symbol>
        <var_symbol>%s</var_symbol>
        <spec_symbol>%s</spec_symbol>
        <price>%s</price>
        <type>%s</type>
        <code>%s</code>
        <status>%s</status>
        <memo>%s</memo>
        <date>%s</date>
        <crtime>%s</crtime>
        <name>%s</name>
      </item>'''


def _escape_data_list(data_list):
    return tuple(escape(item if isinstance(item, types.StringTypes) else str(item)) for item in data_list)


def render_template_head(data_list):
    return _template_head % _escape_data_list(data_list)

def render_template_tail():
    return _template_tail

def render_template_item(data_list):
    return _template_item % _escape_data_list(data_list)


proc_ebanka_csv.py
#! /usr/bin/env python2
#
# Copyright (C) 2009-2019  CZ.NIC, z. s. p. o.
#
# This file is part of FRED.
#
# FRED is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# FRED is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with FRED.  If not, see <https://www.gnu.org/licenses/>.

# vyznamy jednotlivych sloupcu v csv souboru:
# 0, 1 - datumy
# 2    - predepsana castka
# 3    - mena
# 4    - prevedena castka, ta se muze zmenit, napr. pri zmene stavu z Nerealizovano na Zrealizovano
# 5    - timestamp (datum je shodne s datumem ze sloupce 0
# 6    - cislo uctu z nejz prisla platba
# 7    - kod banky z nejz prisla platba
# 8    - cislo uctu na nejz prisla platba
# 9    - kod bankdy na nejz prisla platba
# 10   - variabilni symbol
# 11   - konstantni symbol
# 12   - poznamka
# 13   - stav
#         0 - Nezrealizovano
#         1 - Castecne realizovano
#         2 - Zrealizovano
#         3 - Suspendovano
#         4 - Ukonceno
#         5 - Cekani na clearing
# 14   - jmeno uctu protistrany
# 15   - identifikacni cislo

import sys
import csv
from datetime import datetime

from fred_transproc.template import render_template_head, render_template_tail, render_template_item

status_mapping = {
    0: 3,
    1: 2,
    2: 1,
    3: 4,
    4: 5,
    5: 6,
}

def error(msg):
    sys.stderr.write('Invalid format: ' + msg + '\n')
    sys.exit(1)

if __name__ == '__main__':
    source_lines = sys.stdin.readlines()
    reader = csv.reader(source_lines, delimiter=";")
    first_time = True
    for row in reader:
        if row == [' ']:
            # if no bank transaction, raifaisen bank returns file with one space :-(
            # we don't want to see this as error so just quit silently
            sys.exit(0)
        if len(row) != 16:
            error('Bad number of columns, text in source is:\n' + ''.join(source_lines))
        if first_time:
            first_time = False
            own_account_number = row[8].strip()
            own_account_bank_code = row[9].strip()
            number = ""
            date = ""
            balance = ""
            old_date = ""
            old_balance = ""
            credit = ""
            debet = ""
            print render_template_head([own_account_number, own_account_bank_code, number, date, balance,
                                        old_date, old_balance, credit, debet])
        ident = row[15].strip()
        account_number = row[6].strip()
        bank_code = row[7].strip()
        const_symbol = row[11].strip()
        var_symbol = row[10].strip()
        spec_symbol = ""
        price = row[4].strip().replace(',', '.').replace(' ', '')
        currency = row[3]
        if currency != 'CZK':
            # transfers which are not in CZK are processed throught
            # raiffeisen TXT reports
            # we set price to zero here in order to import the payment
            # and be sure that it will not be processed by backend
            # till correct price in CZK is known
            price = 0
        type = "1"  # not decided (not processed)

        # status
        #    * 1-Realized (only this should be further processed)
        #    * 2-Partially realized
        #    * 3-Not realized
        #    * 4-Suspended
        #    * 5-Ended
        #    * 6-Waiting for clearing
        status = status_mapping[int(row[13].strip())]
        # import only realized payments
        if status != 1:
            continue

        # all transfers in CSV file are deposit from registrars
        code = "1"
        memo = row[12].strip()[:64]
        date = datetime.strptime(row[5].strip(), "%d.%m.%Y %H:%M:%S").date().isoformat()
        crtime = ""
        name = row[14].strip()
        print render_template_item([ident, account_number, bank_code, const_symbol,
                                    var_symbol, spec_symbol, price, type, code, status, memo,
                                    date, crtime, name])
    if first_time == False: # there was at least one row in input csv
        print render_template_tail()
